terraform/modules/s3_bucket/log_encrypted_bucket/log_encrypted_bucket.tf

```
@@ -0,0 +1,71 @@
+resource "aws_s3_bucket" "log_encrypted_bucket" {
+  bucket        = var.bucket
+  force_destroy = var.force_destroy
+}
+
+
+resource "aws_s3_bucket_server_side_encryption_configuration" "log_encrypted_bucket_sse_config" {
+  bucket = aws_s3_bucket.log_encrypted_bucket.id
+  rule {
+    apply_server_side_encryption_by_default {
+      sse_algorithm = "AES256"
+    }
+  }
+}
+
+resource "aws_s3_bucket_versioning" "log_encrypted_bucket_versioning" {
+  count = var.versioning ? 1 : 0
+  bucket = aws_s3_bucket.log_encrypted_bucket.id
+  versioning_configuration {
+    status = "Enabled"
+  }
+}
+resource "aws_s3_bucket_acl" "log_encrypted_bucket_acl" {
+  bucket = aws_s3_bucket.log_encrypted_bucket.id
+  acl    = var.acl
+}
+
+resource "aws_s3_bucket_lifecycle_configuration" "log_encrypted_bucket_lifecycle" {
+  bucket = aws_s3_bucket.log_encrypted_bucket.id
+  rule {
+    id = "log-rule"
+    filter {
+      prefix  = ""
+    }
+    status = var.expiration_days == 0 ? "Disabled" : "Enabled"
+    transition {
+      days          = 365
+      storage_class = "ONEZONE_IA"
+    }
+    expiration {
+      # Hack: Set expiration days to 30 if unset; objects won't actually be expired because the rule will be disabled
+      # See https://github.com/terraform-providers/terraform-provider-aws/issues/1402
+      days = var.expiration_days == 0 ? 30 : var.expiration_days
+    }
```

What actually disables the rule if it's set to expire in 30 days?

Also, did you see this comment in that thread? https://github.com/hashicorp/terraform-provider-aws/issues/1402#issuecomment-514480313

I think it might be preferable to use `for_each` to make the block truly conditional, so that we can guarantee that the expiration lifecycle rule is only added when we want.

I think you can use the `for_each` approach if you add this to the file somewhere:

```
locals {
  expiration_enabled = var.expiration_days == 0 ? [] : ["enabled"];
}
```

And then change this block like so:

```suggestion
  dynamic "expiration" {
      for_each = local.expiration_enabled
      content {
        days = var.expiration_days
      }
    }
```

---

I just copied this code from the `encrypted_bucket` terraform code, and added the section for transition. I can try updating this with dynamic blocks, but right now the other buckets seem to be working.

---

Added a comment showing where the rule gets disabled. I think adding an extra variable and a dynamic block would make this more difficult to understand and read.